### ISSUE TITLE
Topology help layers and working with cetroids

### ISSUE THREAD

**Issue by signedav** [opengisch/QgisModelBaker#689]
**Request from a customer:**

Wenn im Interlis-Modell auf einer Klasse X die AREA Constraint definiert ist, soll es möglich sein, dass beim Schemaimport (Generate Project) automatisch folgende 2 Hilfslayer generiert werden:
- Edges (Lines) für diese Klasse X
- Centroids (Points) für diese Klasse X
Die Edges enthalten keine im Interlis-Modell definierten Attribute
Die Centroids enthalten alle im Interlis-Modell definierten Attribute dieser Klasse X

Nice to have: Bei einem Datenimport werden aus den Polygonen dieser Klasse automatisch auch die beiden Hilfslayer abgefüllt. Die Edges würden im Sinne des Saga-Tools "Polygons to Edges and Nodes" oder gleich wie wenn man Polygone in einen Linienlayer kopiert gebildet. Pro Polygon würde auch ein Centroid (Punkt) erstellt. Dieser müsste zwingend innerhalb des Polygons gesetzt werden, ist streng genommen also kein "Zentroid". Vielleicht gibt es auch ein passenderes Wort. GEONIS für ArcGIS verwendet einfach dieses Wort.

Zwingendes Feature wäre aber dann, dass bei jeder Validierung die Klasse X komplett überschrieben wird, und zwar mit den Polygonen, die entstehen, wenn man aus den Edges Polygone bildet (entsprechendes Tool wäre "Polygonize"), und mit den Attributen aus dem Centroid, der innerhalb der umrahmenden Edges liegt. In diesem Validierungsschritt würde gleichzeitig getestet, ob jedes durch Polygonize gebildetete Polygon EXAKT ein Centroid enthält (Null Centroids würde eine Fehlermeldung generieren, ebenso mehr als ein Centroid).

Die Endpunkte der einzelnen Linien im Edges-Layer müssen nicht aufeinanderliegen, der Layer darf also Dangles enthalten.

Diese Funktionaltiätserweiterung würde das systematische Erfassen von Klassen mit AREA Constraints erlauben, ohne immer Gefahr zu laufen, Lücken oder Überlagerungen zu generieren.

<img width="404" alt="GW_00001" src="https://user-images.githubusercontent.com/28384354/165965653-a083573d-1aa8-40fd-b687-19eb58e1a8cd.png">

Bildbeispiel: Layer Edges. Dunkelgrün: Liegenschaftskanten; Rot/Lila: Zonenabgrenzungen (Schnittlinien mit Dangles)

**Comment by signedav** [comment]
@romefi Do you know if someone is working like that and could provide some information how QGIS / Model Baker users handled such use cases? I know about some ili2db arguments like `--skipPolygonBuilding` but those seems more to be for fixing invalid data.

**Comment by webrian** [comment]
@romefi Yes, you're right. We use the INTERLIS 1 model [DM16NplBE](https://www.agi.dij.be.ch/de/start/geoportal/datenmodelle/beschreibung-kantonaler-datenmodelle.html) with "--skipPolygonBuilding" and "--keepAreaRef".
A table like

```
TABLE GN_Grundzone =
        ZonentypRef: -> GN_Zonentyp; !!//TS:Geometrie.Eigentumsbeschraenkung//
        Genehmigung: OPTIONAL DATE;
        Inkraftsetzung: OPTIONAL DATE;
        Laermempfindlichkeit: OPTIONAL Laermempfindlichkeit;
        Beschrieb: OPTIONAL TEXT*80;
        Planbeschriftung: OPTIONAL TEXT*150; !! fuer Beschriftungen auf dem Plan 
        Geometrie: AREA WITH (STRAIGHTS, ARCS) VERTEX LKoord 
            WITHOUT OVERLAPS > 0.050; !!//TS:Geometrie.Flaeche//
        Qualitaet: Qualitaetsstandard;
        Bemerkung: OPTIONAL TEXT*80;
        Flaechenmass: DIM2 0 999999999;  !! aus Koordinaten gerechnete Flaeche
    NO IDENT    
    END GN_Grundzone;
```

results in two tables in the database, one table with all attributes and the centroid:

```
CREATE TABLE public.gn_grundzone (
  t_id int8 NOT NULL DEFAULT nextval('t_ili2db_seq'::regclass),
  t_ili_tid varchar(200) NULL,
  genehmigung date NULL,
  inkraftsetzung date NULL,
  ...
  geometrie_ref geometry(point, 2056) NOT NULL,
  ...
);
```
and a second one with the boundaries only:

```
CREATE TABLE public.gn_grundzone_geometrie (
  t_id int8 NOT NULL DEFAULT nextval('t_ili2db_seq'::regclass),
  t_ili_tid varchar(200) NULL,
  "_geom" geometry(compoundcurve, 2056) NOT NULL,
  CONSTRAINT gn_grundzone_geometrie_pkey PRIMARY KEY (t_id)
);
```

In QGIS the boundaries and the centroids:
![image](https://user-images.githubusercontent.com/980073/166163276-98f3db31-c9de-43c9-9321-d7fb525af23a.png)

Using a PostGIS view the areas can be visualized:
![image](https://user-images.githubusercontent.com/980073/166163297-eb7b37ff-9d93-4d32-bcd2-1e19440d2ab9.png)

But as far as I know this won't work for INTERLIS 2 models.

**Comment by romefi** [comment]
Thanks @webrian for the very nice example. So it works in INTERLIS 1, but not in INTERLIS 2.

@signedav I think it would be a bunch of work to make this behaviour also available also in INTERLIS 2 (ili2db + Model Baker). Could you doublecheck with th customer what he thinks about it and if he wants to have that with INTERLIS 2, if he's maye willing to contribute something? Thank you.

**Comment by kingemir** [comment]
Hi, here's "the customer". 
We use Interlis 2. We would want this feature for Interlis 2.
